### PR TITLE
add ecdsa signature

### LINKED ISSUE #28: ecdsa
The `ecdsa` builtin is a `SignatureBuiltin`, see for example https://github.com/starkware-libs/cairo-lang/blob/efa9648f57568aad8f8a13fbf027d2de7c63c2c0/src/starkware/cairo/common/signature.cairo#L10

### PR DESCRIPTION
**Pull request by AntoineFONDEUR**
Add ecdsa signature

Time spent on this PR: 3 hours

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

No ecdsa.

Resolves #28 

## What is the new behavior?

Added two functions : sign and check.

### CODE REVIEW

File: src/builtins/ecdsa/ecdsa.ts
Comment (ClementWalter) [inline src/builtins/ecdsa/ecdsa.ts:1]: be careful with the case! in ts, let's only use `camelCase`

File: src/builtins/ecdsa/ecdsa.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  }
+  var s: bigint = modMul(
+    modInv(k, ORDER),
+    modAdd(message, modMul(r, private_key, ORDER), ORDER),
+    ORDER,
+  );
+  return { r: r, s: s } as SignatureType;
+}
+
+function verify_sig(
```
Comment (ClementWalter) [inline src/builtins/ecdsa/ecdsa.ts:16]: no abbreviation please, so `verify_signature`

File: src/builtins/ecdsa/ecdsa.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  public_key: AffinePoint,
+) {
+  var s_inv: bigint = modInv(signature.s, ORDER);
+  var u: bigint = modMul(message, s_inv, ORDER);
+  var v: bigint = modMul(signature.r, s_inv, ORDER);
+  var c: AffinePoint = ec_add(ec_mul(u, GENERATOR), ec_mul(v, public_key));
+  return c.x === signature.r;
+}
+
+function getRandomBigInt(n: bigint): bigint {
```
Comment (ClementWalter) [inline src/builtins/ecdsa/ecdsa.ts:28]: I'd go for something string based instead and avoid the "random" for loop, similar to https://stackoverflow.com/a/49232990/4444546

```
target = n.toString(16).length
// generate random hex string with the appropriate length
return BitIng(random) % n
```

File: src/tests.ts [hunk 3]
```diff
@@ -93,6 +93,14 @@ function testRunner() {
     t.equal(hash, expectedHash);
   });
 
+  test("ecdsa", (t) => {
```
Comment (ClementWalter) [inline src/tests.ts:96]: can we add test cases with bigger values? (closer to the order of the order)

File: src/vm/constants.ts
Comment (ClementWalter) [inline src/vm/constants.ts:1]: can we put links to the cairo vm python where these values come from?

**Review by ClementWalter — CHANGES_REQUESTED** [review]
mainly check the case and add some test cases

File: src/vm/constants.ts
Comment (AntoineFONDEUR) [reply src/vm/constants.ts:1]: All values from examples come from a sagemath notebook I created.
The generator comes from the starknet documentation and the order was calculated with sagemath.

File: src/builtins/ecdsa/ecdsa.ts (hunk 2 above)
Comment (AntoineFONDEUR) [reply src/builtins/ecdsa/ecdsa.ts:28]: Tell me if you had smth else in mind

File: src/builtins/ecdsa/ecdsa.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
+function signMessage(message: bigint, private_key: bigint): SignatureType {
+  var k: bigint = getRandomBigInt(ORDER - BigInt(1)) + BigInt(1);
+  var r_point: AffinePoint = ec_mul(k, GENERATOR);
```
Comment (ClementWalter) [inline src/builtins/ecdsa/ecdsa.ts:3]: case

File: src/builtins/ecdsa/ecdsa.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    modAdd(message, modMul(r, private_key, ORDER), ORDER),
+    ORDER,
+  );
+  return { r: r, s: s } as SignatureType;
+}
+
+function verifySignature(
+  signature: SignatureType,
+  message: bigint,
+  public_key: AffinePoint,
```
Comment (ClementWalter) [inline src/builtins/ecdsa/ecdsa.ts:19]: case

File: src/builtins/ecdsa/ecdsa.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+) {
+  var s_inv: bigint = modInv(signature.s, ORDER);
```
Comment (ClementWalter) [inline src/builtins/ecdsa/ecdsa.ts:21]: case

**Review by ClementWalter — APPROVED** [review]

**Pull request merged**